pkg/dockerregistry/server/writelimiter/blobstore.go

```diff
@@ -0,0 +1,119 @@
+package writelimiter
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	"github.com/openshift/origin/pkg/dockerregistry/server/types"
+)
+
+// ErrResourcesExhausted is returned when no more BlobWriters can be created
+// because of the limit.
+var ErrResourcesExhausted = errors.New("writers limit is reached")
+
+type writerBouncer struct {
+	sem  *semaphore
+	once sync.Once
+}
+
+func newWriterBouncer(sem *semaphore) (*writerBouncer, error) {
+	if !sem.TryDown() {
+		return nil, ErrResourcesExhausted
+	}
+	return &writerBouncer{
+		sem: sem,
+	}, nil
+}
+
+func (g *writerBouncer) Release() {
+	g.once.Do(g.sem.Up)
+}
+
+type blobWriter struct {
+	distribution.BlobWriter
+	bouncer *writerBouncer
+}
+
+var _ distribution.BlobWriter = blobWriter{}
+
+func (bw blobWriter) Close() error {
+	bw.bouncer.Release()
+	return bw.BlobWriter.Close()
+}
+
+func (bw blobWriter) Commit(ctx context.Context, desc distribution.Descriptor) (distribution.Descriptor, error) {
+	bw.bouncer.Release()
+	return bw.BlobWriter.Commit(ctx, desc)
+}
+
+func (bw blobWriter) Cancel(ctx context.Context) error {
+	bw.bouncer.Release()
+	return bw.BlobWriter.Cancel(ctx)
+}
+
+type blobStore struct {
+	distribution.BlobStore
+	sem *semaphore
+}
+
+var _ distribution.BlobStore = blobStore{}
+
+func (bs blobStore) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
+	bouncer, err := newWriterBouncer(bs.sem)
```

Comment:
If minimum chunk size is 5mb, that means 200 uploads per gig.  We can probably conservatively get to 5-10gb on this machine, which is up to 2k layers at once.  That's probably enough to get us to the next step.